### ISSUE TITLE
ImagePush returns HTTP 200 in spite of failure

### ISSUE THREAD

**Issue by cowwoc** [moby/moby#49148]
### Description

DigitalOcean's docker container has a free tier that is limited to 500MB. If you attempting to push an image that surpasses this limit then the local Docker Engine returns an `errorDetail` containing `failed to authorize: failed to fetch oauth token: unexpected status from GET request to https://api.digitalocean.com/v2/registry/auth?scope=repository%3Alicensed-app%2Fshared%3Apull&scope=repository%3Alicensed-app%2Fshared%3Apull%2Cpush&service=registry.digitalocean.com: 401 Unauthorized`

That's fine on its own, but then https://docs.docker.com/reference/api/engine/version/v1.47/#tag/Image/operation/ImagePush returns `HTTP 200` which is clearly incorrect.

### Reproduce

1. Create an account on digitalocean
2. Create a docker container registry using the free tier
3. Keep on pushing images to it until you surpass the 500MB limit
4. Notice that the local Docker Engine will return a `errorDetail` but then return HTTP 200 as a response code.

### Expected behavior

The Engine should return an HTTP code that indicates some sort of failure.

### docker version

```bash
Client:
 Version:           27.4.0
 API version:       1.47
 Go version:        go1.22.10
 Git commit:        bde2b89
 Built:             Sat Dec  7 10:40:21 2024
 OS/Arch:           windows/amd64
 Context:           desktop-linux

Server: Docker Desktop 4.37.0 (178034)
 Engine:
  Version:          27.4.0
  API version:      1.47 (minimum version 1.24)
  Go version:       go1.22.10
  Git commit:       92a8393
  Built:            Sat Dec  7 10:38:57 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.21
  GitCommit:        472731909fa34bd7bc9c087e4c27943f9835f111
 runc:
  Version:          1.1.13
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    27.4.0
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  ai: Ask Gordon - Docker Agent (Docker Inc.)
    Version:  v0.5.1
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-ai.exe
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.19.2-desktop.1
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-buildx.exe
  compose: Docker Compose (Docker Inc.)
    Version:  v2.31.0-desktop.2
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-compose.exe
  debug: Get a shell into any image or container (Docker Inc.)
    Version:  0.0.37
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-debug.exe
  desktop: Docker Desktop commands (Beta) (Docker Inc.)
    Version:  v0.1.0
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-desktop.exe
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.2
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-dev.exe
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.27
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-extension.exe
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.5
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-feedback.exe
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.4.0
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-init.exe
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-sbom.exe
  scout: Docker Scout (Docker Inc.)
    Version:  v1.15.1
    Path:     C:\Users\Gili\.docker\cli-plugins\docker-scout.exe

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 60
 Server Version: 27.4.0
 Storage Driver: overlayfs
  driver-type: io.containerd.snapshotter.v1
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 CDI spec directories:
  /etc/cdi
  /var/run/cdi
 Swarm: inactive
 Runtimes: nvidia runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 472731909fa34bd7bc9c087e4c27943f9835f111
 runc version: v1.1.13-0-g58aa920
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
 Kernel Version: 5.15.153.1-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 31.31GiB
 Name: docker-desktop
 ID: 9cd2a19c-6a32-4004-9d22-c6f6db240e52
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Labels:
  com.docker.desktop.address=npipe://\\.\pipe\docker_cli
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
WARNING: daemon is not using the default seccomp profile
```

**Comment by thaJeztah** [comment]
Thanks for reporting.

I think this is the expected behavior, not a bug, but admittedly can be confusing (and somewhat more involved) to handle.

The endpoints used for pushing (and pulling) images provide a streaming response; the 200 status returned when making the request indicates that the initial request was received, and correct, but indicates the _start_ of the push / pull. The pull itself involves many requests made to the registry and during those requests, errors may occur, including a failure to access a resource due to authentication or otherwise. Unfortunately it's not possible to detect such cases ahead of time and to adjust the initial status returned based on that, because those errors are the result of follow-up requests made by the daemon and detecting the error involves _making_ the request (some of which happening in parallel).

So to handle these situations, the client must process the streaming response returned, and handle errors that are returned as part of that response.

There's definitely things to be improved here, both in the implementation as well as _documenting_ some of  this logic; I created a tracking ticket for some of this a while back, but contributions (e.g. to improve documentation) are definitely welcome;

- https://github.com/moby/moby/issues/46528

**Comment by thaJeztah** [comment]
@vvoland I see you added the containerd-integration  label; am I mistaken  and is there some bug in this area, or did you add this for a different reason?

**Comment by cowwoc** [comment]
@thaJeztah makes sense... I agree that improving the documentation would help a lot in this case.

In terms of making the server response more intuitive, you could make long-running processes return HTTP 202 ("Accepted") with a `Location` header that points to a status monitor endpoint. The latter would contain the cumulative state of the process, meaning the returned JSON object could contain a `stream` array that is populated with output as it arrives over time. You would have a separate property to indicate the status of the operation (e.g. pending, success, failure) and on failure you could provide an additional property containing the error message (or the last `stream` entry could contain it).

You could also return a `Retry-After` header to indicate how often you'd like clients to poll for a status update.

In any case... that's one approach.

**Comment by vvoland** [comment]
I saw the docker info mentioned the containerd integration so just wanted to mark this as possibly related. Removed it now.

**Comment by thaJeztah** [comment]
> I saw the docker info mentioned the containerd integration so just wanted to mark this as possibly related. Removed it now.

Thanks! Just making sure wasn't tapering over some actual regression / issue :heart:

> In terms of making the server response more intuitive, you could make long-running processes return HTTP 202 ("Accepted") with a `Location` header that points to a status monitor endpoint. The latter would contain the cumulative state of the process, meaning the returned JSON object could contain a `stream` array that is populated with output as it arrives over time. You would have a separate property to indicate the status of the operation (e.g. pending, success, failure) and on failure you could provide an additional property containing the error message (or the last `stream` entry could contain it).
>
> You could also return a `Retry-After` header to indicate how often you'd like clients to poll for a status update.
>
> In any case... that's one approach.

Thanks for those suggestions; perhaps worth re-posting them on https://github.com/moby/moby/issues/46528 to keep things central.

There's quite some things we need to consider when changing those APIs; there have been some discussions about options to make them more declarative, as well as requests for being able to monitor progress (i.e., being able to see all pulls currently in progress). Those are things to consider, but must be done carefully. Such changes may even require introduction new endpoints, or selectable behavior.

We know there's millions of existing installations, and many integrations; while we do version the API, not all integrations may properly use versions into account, and (e.g.) changing a `200` status to `202` even when made API-version dependent could easily break such integrations. For that reason, we try to avoid making "papercut" breaking changes in some cases, and for those it may be better to have a "big bang" redesign instead.

**Comment by thaJeztah** [comment]
Based on the above, I think the direct question was answered, and wee have at least https://github.com/moby/moby/issues/46528 in place to track improving these endpoints; let me close this one to prevent the discussion around  those improvements from diverging (there's likely other tickets as well to de-duplicate 🙈)
